### PR TITLE
add preprocessing step

### PR DESCRIPTION
**Pull request by unie31**

### CODE REVIEW

File: sim/preprocessing.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    return funcs
+
+def preprocess_stands(stands: List[ForestStand], operations: list) -> List[ForestStand]:
+    """Applies the given preprocessing operations to the given stands
+    :returns: preprocessed stands
+    """
+    funcs = __get_preprocessing_funcs(operations)
+    for func in funcs:
+        stands = func(stands)
+    return stands
```
Comment (jpy-luke) [inline sim/preprocessing.py:31]: This is essentially the same function as evalute_sequence in runners.py. No need for this.

**Review by jpy-luke — COMMENTED** [review]
Left comments. Let's talk about this over a call. I need to open up the concept of partially applied, prepared functions in our project.

File: forestry/preprocessing.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+def exclude_sapling_trees(stands: List[ForestStand], **operation_params) -> List[ForestStand]:
+    for stand in stands:
+        #is there benefit from using filter here since it returns a generator which should be read from straight away? 
+        stand.reference_trees = filter(lambda t: (t if t.sapling is False else None), stand.reference_trees)
+    return stands
+
+def exclude_empty_stands(stands: List[ForestStand], **operation_params)-> List[ForestStand]:
+    #is there benefit from using filter here since it returns a generator which should be read from straight away? 
+    stands = filter(lambda s: (s if len(s.reference_trees) > 0 else None), stands)
+    return stands
```
Comment (jpy-luke) [inline forestry/preprocessing.py:13]: You are returning the filter iterator from these. Need to return list.

File: sim/generators.py [hunk 3]
```diff
@@ -178,3 +189,12 @@ def generate_processor(operation_lookup, operation_params, operation_tag, run_co
         this_run_constraints,
         **this_operation_params)
     return result
+
+def get_preprocessing_funcs(operations: list, operation_params: dict, operation_lookup: dict) -> List[Callable]:
```
Comment (jpy-luke) [inline sim/generators.py:193]: This function is not needed anymore, is it?

File: forestry/preprocessing.py (hunk 2 above)
Comment (unie31) [reply forestry/preprocessing.py:13]: added casting to list in commit da01d8d, not sure why the change is not visible in that snippet, although it is in the file

File: sim/generators.py (hunk 3 above)
Comment (unie31) [reply sim/generators.py:193]: no, slipped through!

**Pull request closed without merging**